### PR TITLE
tt: add instances command

### LINKED ISSUE #249: Add the ability to get a list of enabled applications (instances)
When you go into an environment it's very convenient to be able to get a list of all the applications in it. We need to give users this possibility.

### PR DESCRIPTION
**Pull request by better0fdead**
Added instances command. It prints a list of enabled applications.

Changelog and Readme updated.

Closes #249

### CODE REVIEW

File: test/integration/list/test_list.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import os
+import re
+import shutil
+import subprocess
+import tempfile
+
+
+def test_list_enabled_apps(tt_cmd):
```
Comment (psergee) [inline test/integration/list/test_list.py:8]: Please, add the following tests:
- `instances_enabled: .` in tt env config. There are two scenarios for this:
    - . is an application. If it is not
    - . is not an app.

- `instances_enabled` points to non-existing directory.
- `instances_enabled` contains a directory, which is not an application (contains any .lua file, but not init.lua)

File: cli/list/list.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		return fmt.Errorf("can't collect an application list: %s", err)
+	}
+
+	if len(appList) == 0 {
+		fmt.Println("there is no enabled applications")
+	}
+
+	fmt.Println("List of enabled applications:")
+	for _, inst := range appList {
+		log.Infof("%s", strings.TrimSuffix(inst.Name, ".lua"))
```
Comment (psergee) [inline cli/list/list.go:33]: Why not fmt.Println?

File: cli/cmd/list.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/list"
+	"github.com/tarantool/tt/cli/modules"
+)
+
+// NewListCmd creates list command.
+func NewListCmd() *cobra.Command {
+	var listCmd = &cobra.Command{
+		Use:   "list",
+		Short: "Show enabled applications.",
```
Comment (0x501D) [inline cli/cmd/list.go:14]: There is no dot in the help of other commands.

File: test/integration/list/test_list.py (hunk 1 above)
Comment (better0fdead) [reply test/integration/list/test_list.py:10]: added

File: cli/list/list.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+	fmt.Printf("Instances enabled directory: %s\n", instEnabledPath)
+
+	appList, err := util.CollectAppList(cmdCtx.Cli.ConfigDir, instEnabledPath, false)
+	if err != nil {
+		return fmt.Errorf("can't collect an application list: %s", err)
+	}
+
+	if len(appList) == 0 {
+		fmt.Println("there is no enabled applications")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/list/list.go]: fixed

File: cli/list/list.go (hunk 2 above)
Comment (better0fdead) [reply cli/list/list.go]: changed

File: cli/cmd/list.go (hunk 3 above)
Comment (better0fdead) [reply cli/cmd/list.go]: removed

**Review by 0x501D — APPROVED** [review]

File: test/integration/list/test_list.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            cwd=test_app_path,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            text=True
+        )
+        start_output = instance_process.stdout.read()
+        assert re.search("test_app", start_output)
+
+
+def test_list_dot_direcory_with_lua_file(tt_cmd):
```
Comment (psergee) [inline test/integration/list/test_list.py:89]:
```suggestion
def test_list_dot_directory_with_lua_file(tt_cmd):
```

File: test/integration/list/test_list.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            cwd=test_app_path,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            text=True
+        )
+        start_output = instance_process.stdout.read()
+        assert re.search("there are no enabled applications", start_output)
+
+
+def test_list_missing_direcory(tt_cmd):
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/list/test_list.py]: fixed

File: test/integration/list/test_list.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            cwd=test_app_path,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            text=True
+        )
+        start_output = instance_process.stdout.read()
+        assert re.search("instances enabled directory doesn't exist", start_output)
+
+
+def test_list_dot_direcory_with_app(tt_cmd):
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/list/test_list.py]: fixed

File: test/integration/list/test_list.py (hunk 5 above)
Comment (better0fdead) [reply test/integration/list/test_list.py]: fixed

**Review by psergee — APPROVED** [review]

**Comment by LeonidVas** [comment]
Hi. Thank you for the patch.
I have a several comments:
* Maybe better to output the relative path to the directory?
* Would it be better to make the list output as in `tt status` only without the statuses? Or somehow else highlight multi instances application. We can discuss in a voice
```
┌─[leonid@vasya-hp]─[~/work/mail/tasks/demo/test_env_3]
└──╼ tt list
Instances enabled directory: /home/leonid/work/mail/tasks/demo/test_env_3/instances.enabled
List of enabled applications:
        lyapunov
        myapp
        test
```
```
┌─[leonid@vasya-hp]─[~/work/mail/tasks/demo/test_env_3]
└──╼ tt status
   • lyapunov: NOT RUNNING
   • myapp:s1-replica: NOT RUNNING
   • myapp:s2-master: NOT RUNNING
   • myapp:s2-replica: NOT RUNNING
   • myapp:stateboard: NOT RUNNING
   • myapp:router: NOT RUNNING
   • myapp:s1-master: NOT RUNNING
   • test: RUNNING. PID: 295358.
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 
 - Follow-up message support for application templates.
+- ``tt list`` command. It prints a list of enabled applications.
```
Comment (LeonidVas) [inline CHANGELOG.md:13]: `tt list`` command to print a list of enabled applications.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi. Thank you for the patch.
I have a several comments:
* Maybe better to output the relative path to the directory?
* Would it be better to make the list output as in `tt status` only without the statuses? Or somehow else highlight multi instances application. We can discuss in a voice
```
┌─[leonid@vasya-hp]─[~/work/mail/tasks/demo/test_env_3]
└──╼ tt list
Instances enabled directory: /home/leonid/work/mail/tasks/demo/test_env_3/instances.enabled
List of enabled applications:
        lyapunov
        myapp
        test
```
```
┌─[leonid@vasya-hp]─[~/work/mail/tasks/demo/test_env_3]
└──╼ tt status
   • lyapunov: NOT RUNNING
   • myapp:s1-replica: NOT RUNNING
   • myapp:s2-master: NOT RUNNING
   • myapp:s2-replica: NOT RUNNING
   • myapp:stateboard: NOT RUNNING
   • myapp:router: NOT RUNNING
   • myapp:s1-master: NOT RUNNING
   • test: RUNNING. PID: 295358.
```

File: CHANGELOG.md (hunk 8 above)
Comment (better0fdead) [reply CHANGELOG.md]: Changed

File: cli/running/running.go [hunk 9]
```diff
@@ -278,7 +278,7 @@ func getInstancesFromYML(dirPath string, selectedInstName string) ([]InstanceCtx
 }
 
 // collectInstances searches all instances available in application.
```
Comment (LeonidVas) [inline cli/running/running.go:280]: `CollectInstances` in comment.

File: cli/util/util.go [hunk 10]
```diff
@@ -920,7 +920,7 @@ type AppListEntry struct {
 }
 
 // CollectAppList collects all the supposed applications in passed appsPath directory.
-func CollectAppList(baseDir string, appsPath string) ([]AppListEntry, error) {
+func CollectAppList(baseDir string, appsPath string, verbose bool) ([]AppListEntry, error) {
```
Comment (LeonidVas) [inline cli/util/util.go:923]: Why do you need this change?
Move changes to the separate commit (the change in many places makes it difficult to see the meaningful code).

File: cli/running/running.go [hunk 11]
```diff
@@ -278,7 +278,7 @@ func getInstancesFromYML(dirPath string, selectedInstName string) ([]InstanceCtx
...
-func collectInstances(appName string, appDir string) ([]InstanceCtx, error) {
+func CollectInstances(appName string, appDir string) ([]InstanceCtx, error) {
```
Comment (LeonidVas) [inline cli/running/running.go:281]: Since you use this function in other modules, it makes sense to move it to some common place.

**Review by LeonidVas — CHANGES_REQUESTED** [review]

File: test/integration/list/test_list.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        assert re.search("instances enabled directory doesn't exist", start_output)
+
+
+def test_list_dot_directory_with_app(tt_cmd):
+    with tempfile.TemporaryDirectory() as tmpdir:
+        test_app_path_src = os.path.join(os.path.dirname(__file__), "test_app")
+        test_app_path = os.path.join(tmpdir, "test_app")
+        shutil.copytree(test_app_path_src, test_app_path)
+        test_app_path = os.path.join(tmpdir)
+        config_path = os.path.join(test_app_path, "tarantool.yaml")
```
Comment (psergee) [inline test/integration/list/test_list.py:73]: Config name is tt.yaml now. Update it in all places.

File: test/integration/list/test_list.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        start_cmd = [tt_cmd, "list"]
+        instance_process = subprocess.Popen(
+            start_cmd,
+            cwd=test_app_path,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            text=True
+        )
+        start_output = instance_process.stdout.read()
+        assert re.search("app1", start_output)
```
Comment (psergee) [inline test/integration/list/test_list.py:27]: Missing instances list check.

File: cli/cmd/list.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/spf13/cobra"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/list"
+	"github.com/tarantool/tt/cli/modules"
+)
+
+// NewListCmd creates list command.
+func NewListCmd() *cobra.Command {
+	var listCmd = &cobra.Command{
+		Use:   "list",
```
Comment (psergee) [inline cli/cmd/list.go:13]: Shouldn't it be a sub-command of `tt list`. Soon we will have a new functionality for installed binaries listing. Will this be `tt list` or other command? For example it can be `tt list instances` and `tt list binaries` and more in future. @LeonidVas what do you think?

File: cli/running/running.go (hunk 9 above)
Comment (better0fdead) [reply cli/running/running.go]: Changed

File: cli/running/running.go (hunk 11 above)
Comment (better0fdead) [reply cli/running/running.go:281]: I don't see another place for it. I can not move it to util module because of import cycles. (running->util->running) And if i want to resolve it i would need to move many other structs to util too. I don't think it is reasonable.

File: test/integration/list/test_list.py (hunk 12 above)
Comment (better0fdead) [reply test/integration/list/test_list.py]: changed

File: cli/cmd/list.go (hunk 14 above)
Comment (better0fdead) [reply cli/cmd/list.go:13]: Made it tt list instances.
Maybe tt list apps will be better?

File: test/integration/list/test_list.py (hunk 13 above)
Comment (better0fdead) [reply test/integration/list/test_list.py:27]: I didn't understand your comment. Can you pls clarify it.

File: cli/util/util.go (hunk 10 above)
Comment (better0fdead) [reply cli/util/util.go:923]: Moved. I need it, since CollectAppList prints to the output messages that are not needed for tt list, but is needed for other commands.

File: test/integration/list/test_list.py (hunk 13 above)
Comment (psergee) [reply test/integration/list/test_list.py:27]: As I can see `app1` is a multi-instance app. If yes, need to check that instances names are in output.

File: cli/running/running.go (hunk 11 above)
Comment (psergee) [reply cli/running/running.go:281]: > I don't see another place for it. I can not move it to util module because of import cycles. (running->util->running) And if i want to resolve it i would need to move many other structs to util too. I don't think it is reasonable.

Agree.
@LeonidVas  I think this can be done under separate refactoring task for moving all common stuff from running package to common place. We already have other commands which use running package API: `tt clean` as an example. So, these changes will affect not only the `list` command.

File: test/integration/list/test_list.py (hunk 13 above)
Comment (better0fdead) [reply test/integration/list/test_list.py]: Okey, sure. Added.

**Review by LeonidVas — APPROVED** [review]
LGTM

**Review by psergee — APPROVED** [review]
